README.md
#mesanger


history.txt
Ruslan Sonkin
21.10.2014:17:04:
1

Ruslan Sonkin
14:17:06:
2

Ruslan Sonkin
14:17:07:
3

Ruslan Sonkin
14:17:08:
123


Ruslan Sonkin
21.10.20 14:17:38:
5

Ruslan Sonkin
 14:17:40:
6


Ruslan Sonkin
21.10.2014:18:00:
5

Ruslan Sonkin
21.10.2014:18:04:
6

Ruslan Sonkin
21.10.2014:18:10:
7


Ruslan Sonkin
21.10.20 14:18:39:
7

Ruslan Sonkin
14:18:42:
8


Ruslan Sonkin
25.10.20 10:31:44:
asdfghjkl;jhgfdsfghjkl;lkjhytretyuioklp




chat3.py
# == chat3.py ==
##-- A.B.Glazov -- 29/09/2020 --
##-- чат, отслеживание партнеров
##-- структура сообщения:   mess_nick, dest_nick, mess_data
##--<3>-- структура принятого сообщения:   ip, mess_nick, dest_nick, mess_data, mess_time


##-- библиотеки отрисовки
from tkinter import *
from tkinter import font
from tkinter import ttk
from tkinter import simpledialog

##==<2>== библиотеки для работы с сетью
from socket import *
import threading
from time import *

# == переменные
##-- цвета элементов оформления
root_color = "thistle"  # -- цвет формы
sel_color = "red"

##-- вспомогательные переменные
btn_width = 14  # -- размер кнопок
text_width = 400  # -- ширина поля чата

##--<3>-- списое партнеров (добавил время)
lst_partn = [("all", "255.255.255.255", -1), ("piter", "192.168.1.200", 0)]
partn_tau = 3000  # -- время ожидания партнера в мс

##==<2>== UDP сокет для приема сообщений
HOST_IN = ''
PORT_IN = 4000
BUFSIZE = 1024
SOCKADDR_IN = (HOST_IN, PORT_IN)
uServSock = socket(AF_INET, SOCK_DGRAM)
uServSock.bind(SOCKADDR_IN)

##==<2>==  очередь для приема сообщений и флаг ее занятости
ls_in = []
busy_in = 0
main_tau = 30  # -- период цикла главной функции

##==<2>== UDP  сокет для отправки сообщений
HOST_OUT = '127.0.0.1'
PORT_OUT = 4000
BUFSIZE = 1024
SOCKADDR_OUT = (HOST_OUT, PORT_OUT)
uCliSock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)
uCliSock.setsockopt(SOL_SOCKET, SO_BROADCAST, 1)


# == функции
##--<3>-- функция приема сообщений для потока приема
def work_in():
    global ls_in
    global busy_in
    while True:
        data, addr = uServSock.recvfrom(BUFSIZE)
        loc_data = data.decode('cp1251')
        loc_data = loc_data.strip()

        mess_time = str(get_timems())  # --<3>--
        st = addr[0] + '|' + loc_data + "|" + mess_time  # -- запомнить IP и строку

        while busy_in:
            sleep(0.001)
        busy_in = 2
        ls_in.append(st)
        busy_in = 0

        sleep(0.001)
    uServSock.close()


##---<2>--  отправить сообщение
def send_mess(mess, ip):
    # print( ip )
    st = mess.encode('cp1251')
    sock_addr = (ip, PORT_OUT)
    uCliSock.sendto(st, sock_addr)


##--<3>-- получить время в миллисекундах
def get_timems():
    return int(time() * 1000)


# == создать форму
root = Tk()
root.resizable(width=False, height=False)
dFont = font.Font(family="helvetica", size=12)
stl = ttk.Style()

root.configure(background=root_color)
stl.configure('.', font=dFont, background=root_color, foreground="black")

ttk.Label(root, text='проcтой чат (версия 1)').grid(row=0, column=0, pady=5)

# == левая панель для текстовых сообщений
pnl_left = Frame(root)
pnl_left.grid(row=2, column=0, columnspan=6, rowspan=10, pady=10, padx=10)

##== панель для вывода получекнных сообщений (основная)
pnl_mess = Frame(pnl_left, width=700)
pnl_mess.grid(row=2, column=0)

###== текстовое поле для вывода получекнных сообщений
tbx_mess = Text(pnl_mess, height=30, wrap=WORD, font=dFont)
tbx_mess.pack(side='left', fill='both', expand=1)

###== полоса прокрутки для текстового поля получекнных сообщений
sbr_mess = Scrollbar(pnl_mess)
sbr_mess['command'] = tbx_mess.yview
tbx_mess['yscrollcommand'] = sbr_mess.set
sbr_mess.pack(side='right', fill='y')

##== панель для вывода сообщений
pnl_send = Frame(pnl_left)
pnl_send.grid(row=3, column=0, pady=5)

###== текстовое поле для вывода получекнных сообщений
tbx_send = Text(pnl_send, height=3, wrap=WORD, font=dFont)
tbx_send.pack(side='left', fill='both', expand=1)

###== полоса прокрутки для текстового поля отправляемых сообщений
sbr_send = Scrollbar(pnl_send)
sbr_send['command'] = tbx_send.yview
tbx_send['yscrollcommand'] = sbr_send.set
sbr_send.pack(side='right', fill='y')


###==<2>== обработчик <Enter> для текстового поля отправки
def fnc_tbxsend(event):
    my_nick = var_nick.get()
    if not my_nick:
        my_nick = simpledialog.askstring("замечание", "введите Ваш ник ")
        var_nick.set(my_nick)

    partn_nick = var_partn.get()
    text_send = tbx_send.get(1.0, END).strip()
    # print(text_send)
    if len(text_send) > 0:
        send_mess(my_nick + "|" + partn_nick + "|" + text_send, HOST_OUT)  # --<2>--
        text_out = "\n(  >>>  " + partn_nick + "  ):    " + text_send
        tbx_mess.insert(END, text_out)
        tbx_send.delete(1.0, END)


tbx_send.bind('<Return>', fnc_tbxsend)

# == правая панель для контактов
pnl_right = Frame(root, background=root_color)
pnl_right.grid(row=0, column=8, rowspan=20)

##-- поле вашего Ника
ttk.Label(pnl_right, text='Ваше имя:').grid(row=0, column=0, sticky=W, padx=5)
var_nick = StringVar()
var_nick.set("")
edt_nick = ttk.Entry(pnl_right, width=btn_width, textvariable=var_nick, font=dFont)
edt_nick.grid(row=1, column=0, padx=5, pady=5, sticky=W)

##-- поле ника партнера
ttk.Label(pnl_right, text='Партнер:').grid(row=2, column=0, sticky=W, pady=5)
var_partn = StringVar()
var_partn.set("")
edt_partn = ttk.Entry(pnl_right, width=btn_width, textvariable=var_partn, font=dFont)
edt_partn.grid(row=3, column=0, padx=5, pady=5, sticky=W)

# == панель со списком партнеров
ttk.Label(pnl_right, text='партнеры').grid(row=5, column=0, padx=5, pady=5)
pnl_partn = Frame(pnl_right)
pnl_partn.grid(row=6, column=0, padx=5, pady=5, sticky=N)

##-- список партнеров с прокруткой
lbx_partn = Listbox(pnl_partn, width=btn_width, height=16, font=dFont)
lbx_partn.pack(side="left", fill="y")
sbr_partn = Scrollbar(pnl_partn, orient="vertical")
sbr_partn.pack(side="right", fill="y")

sbr_partn.config(command=lbx_partn.yview)
lbx_partn.config(yscrollcommand=sbr_partn.set)


##--<2>-- обработать двойной клик по элементу списка
def fnc_setpartn(ev):
    global HOST_OUT
    index = lbx_partn.curselection()
    partn = lbx_partn.get(index)
    var_partn.set(partn)
    # print(index, lst_partn)
    HOST_OUT = lst_partn[index[0]][1]
    # print( HOST_OUT)


lbx_partn.bind("<Double-Button-1>", fnc_setpartn)

# == действия при запуске программы
cur_time = get_timems()
lst_partn[1] = ("piter", "192.168.1.200", cur_time)
##-- загрузить список партнеров

for partn in lst_partn:
    lbx_partn.insert(END, partn[0])

##--<2>-- выбрать начального партнера
lbx_partn.selection_set(0)
partn = lbx_partn.get(0)
var_partn.set(partn)
HOST_OUT = lst_partn[0][1]

##--<2>-- создать и запустить поток приема сообщений
tr_in = threading.Thread(target=work_in)
tr_in.daemon = True
tr_in.start()

##--<3>-- запомнить время запуска
time_stamp = get_timems()


##==<3>== главная функция, запускаемая в цикле
def main():
    global ls_in
    global busy_in
    global time_stamp
    global HOST_OUT

    flg_partnchange = 0  # -- сбросить признак изменения списка пратнеров
    my_nick = var_nick.get()
    if len(ls_in) > 0:
        # -- аккуратно извлечь сообщение из очереди
        while busy_in:
            sleep(0.001)
        busy_in = 2
        mess_in = ls_in.pop(0)
        busy_in = 0
        # -- вывести его в поле приема
        lst_mess = mess_in.split("|")

        mess_ip, mess_nick, dest_nick, mess_data, mess_time = lst_mess  # --<3>--
        print(mess_ip, mess_nick, dest_nick, mess_data, mess_time)
        mess_time = int(mess_time)
        if mess_nick != my_nick:

            if mess_data != "%*%":
                text_in = "\n(  " + mess_nick + "  >>>>  ):    " + mess_data
                tbx_mess.insert(END, text_in)

            ##--<3>-- найти партнера в списке по IP
            sel_num = -1
            for num, partn in enumerate(lst_partn):
                if partn[1] == mess_ip and partn[0] == mess_nick:
                    lst_partn[num] = (partn[0], partn[1], mess_time)
                    sel_num = num
                    break
            ##--<3>-- добавить, если не нашли
            if sel_num < 0:
                lst_partn.append((mess_nick, mess_ip, mess_time))
                flg_partnchange = 1

    ##--<3>-- проверить время и удалить выключенные записи
    cur_time = get_timems()
    for num in range(len(lst_partn) - 1, -1, -1):
        if lst_partn[num][2] == -1:  # -- запрещаем удалять all
            continue
        if cur_time > lst_partn[num][2] + 3 * partn_tau:
            del lst_partn[num]
            flg_partnchange = 1

    ##--<3>-- показать изменения списка партнеров
    if flg_partnchange == 1:
        lbx_partn.delete(0, END)
        for partn in lst_partn:
            lbx_partn.insert(END, partn[0])

        lbx_partn.selection_set(0)
        partn = lbx_partn.get(0)
        var_partn.set(partn)
        HOST_OUT = "255.255.255.255"

    ##--<3>-- отправить всем пустое сообшение
    if cur_time - time_stamp >= partn_tau:
        time_stamp = cur_time

        if not my_nick:
            my_nick = simpledialog.askstring("замечание", "введите Ваш ник ")
            var_nick.set(my_nick)

        mess = my_nick + "|all|%*%"
        send_mess(mess, "255.255.255.255")
        sleep(0.005)

    root.after(main_tau, main)


main()

# == запустить обработку событий
root.mainloop()

# ==<2>== закрыть сокеты по завершению программы
uCliSock.close()



functions.py
from socket import *


def send_answer(ip, answer):
    client = socket(AF_INET, SOCK_STREAM)
    try:
        client.connect((ip, 5401))
        client.sendall(answer.encode('utf-8'))
        data = client.recv(1024)
        client_msg = data.decode('utf-8')
        # print(client_msg)
    except:
        client_msg = 'not connected'
        print('not connected')
    finally:
        client.close()

    return client_msg


def send_to_all(msg, user_list):
    '''
    Функция для отправки сообщений всем клиентам подключившимся
    к этому серверу.
    : msg : Отправляемое сообщение (string)
    : user_list : Список подключенных клиентов (list[name, ip])

    Function for sending message to every clients connected
    to this server.
    :param msg: Sending message (string)
    :param user_list: List of connected clients (list[name, ip])
    :return:
    '''
    for element in user_list:
        ip = element[1]
        send_answer(ip, msg)


Ver0.5(GUI_server).py
from tkinter import *
from socket import *
import threading
from time import *
from tkinter import simpledialog


name = "Ruslan Sonkin"
list_in = []                        # очередь для принятых сообщений
busy_in = 0                         # признак занятости очереди



def tools_window_f():
    tools_window = Toplevel()
    tools_window.title('tools')


def options_window_f():
    options_window = Toplevel()
    options_window.title('options')


def clear_history():
    text.config(state=NORMAL)
    text.delete(0.0, END)
    text.config(state=DISABLED)


def send_msg(event):
    text.config(state=NORMAL)
    msg = name + ':\n     ' + entry.get() + '\n\n'
    # text.highlight_pattern(name, "red")
    text.insert(END, msg)
    text.config(state=DISABLED)
    entry.delete(0, END)


server_address = ('localhost', 5400)
def send_2_server(event):
    msg = entry.get()
    entry.delete(0, END)
    # bin_msg = msg.encode('utf-8')             !!!!!!!!!!!!!!!
    bin_msg = bytes(msg, 'UTF-8')

    server_msg = ''

    client = socket(AF_INET, SOCK_STREAM)
    try:
        client.connect(server_address)
        client.sendall(bin_msg)
        data = client.recv(1024)
        server_msg = data.decode('utf-8')
        print(server_msg)
    except:
        server_msg = 'not connected'
        print('not connected')
    finally:
        client.close()

    text.config(state=NORMAL)
    text.insert(END, server_msg + '\n\n')
    text.config(state=DISABLED)


# Аккуратно записать сообщение в очередь
def put_msg(msg: str):
    global list_in
    global busy_in
    while busy_in:
        sleep(0.001)
    busy_in = 1
    list_in.append(msg)
    busy_in = 0


# Функция приема сообщений (один раз за пакет)
def work_in():
    global list_in
    global busy_in

    local_server_address = ('', 5401)
    local_socket = socket(AF_INET, SOCK_STREAM)
    local_socket.bind(local_server_address)
    local_socket.listen(5)

    while True:
        # Ждать и получить сообщение от сервера...
        print('ready')
        connection, address = local_socket.accept()
        print('connection. Address:', address)
        bin_data = connection.recv(1024)
        str_data = bin_data.decode('utf-8')
        print(str_data, "from work_in")
        connection.close()

        # ...записать его в очередь сообщений
        put_msg(str_data)
        
        # задержка потока
        sleep(0.001)


# Настройка потока приема сообщений
threading_in = threading.Thread(target=work_in)
threading_in.daemon = True
threading_in.start()


def display_msg(msg):
    text.config(state=NORMAL)
    text.insert(END, msg + '\n')
    text.yview(END)
    text.config(state=DISABLED)


# Главная функция запускаемая в цикле
def main():
    global list_in
    global busy_in

    if len(list_in) > 0:
        while busy_in:
            sleep(0.001)
        busy_in = 1
        str_in = list_in.pop()
        busy_in = 0

        # print(str_in, 'from main func')
        display_msg(str_in)

    root.after(20, main)    # after(время_задержки_между_повторениями_в_мс, имя_функции)


def registr():
    # global server_address

    def registr_f():
        serv_addr = reg_name_entry.get()


    registr_window = Toplevel(root)
    registr_window.title('Окно регистрации')
    reg_server_address_entry = Entry(registr_window)
    reg_server_address_entry.pack()
    reg_name_entry = Entry(registr_window)
    reg_name_entry.pack()
    reg_btn = Button(text='Зарегистрироваться!', command=registr_f)
    reg_btn.pack()

# Создание интерфейса
root = Tk()
root.geometry('275x400+200+200')
root.title('messenger v0.1')

menu = Menu(root)
root.config(menu=menu)

fileMenu = Menu(menu, tearoff=0)
menu.add_cascade(label='File', menu=fileMenu)

fileMenu.add_command(label='Очистка истории сообщений', command=clear_history)
fileMenu.add_command(label='Выход', command=exit)
menu.add_command(label='Tools', command=tools_window_f)
menu.add_command(label='Options', command=options_window_f)

frame_text = Frame(root, )
frame_text.place(x=5, y=5, relwidth=1, relheight=1, width=-5, height=-30)
text = Text(frame_text, state=DISABLED)
text.place(x=0, y=0, relwidth=1, relheight=1, width=-60, height=0)

btn1 = Button(frame_text, text=1, width=7, height=3)
btn1.pack(anchor=NE, pady=5)
btn2 = Button(frame_text, text=2, width=7, height=3)
btn2.pack(anchor=NE, pady=5)
btn3 = Button(frame_text, text=3, width=7, height=3)
btn3.pack(anchor=NE, pady=5)

frame_entry = Frame(root, )
frame_entry.pack(side=BOTTOM, fill=X)

entry = Entry(frame_entry)
entry.pack(side=LEFT, fill=X, expand=1)

btn = Button(frame_entry, text='send', command=lambda: send_2_server(0))
btn.pack(side=RIGHT, padx=5)

entry.bind('<Return>', send_2_server)

btn1.config(state=DISABLED)
btn2.config(state=DISABLED)
btn3.config(state=DISABLED)

main()

root.mainloop()


Ver0.6(GUI_server).py
from tkinter import *
from socket import *
import threading
from time import *
from tkinter import simpledialog

name = "Ruslan Sonkin"
list_in = []  # очередь для принятых сообщений
busy_in = 0  # признак занятости очереди


server_address = ('localhost', 5400)


# Аккуратно записать сообщение в очередь
def put_msg(msg: str):
    global list_in
    global busy_in
    while busy_in:
        sleep(0.001)
    busy_in = 1
    list_in.append(msg)
    busy_in = 0


# Функция приема сообщений (один раз за пакет)
def work_in():
    global list_in
    global busy_in

    local_server_address = ('', 5401)
    local_socket = socket(AF_INET, SOCK_STREAM)
    local_socket.bind(local_server_address)
    local_socket.listen(5)

    while True:
        # Ждать и получить сообщение от сервера...
        print('ready')
        connection, address = local_socket.accept()
        print('connection. Address:', address)
        bin_data = connection.recv(1024)
        str_data = bin_data.decode('utf-8')
        print(str_data, "from work_in")
        connection.close()

        # ...записать его в очередь сообщений
        put_msg(str_data)

        # задержка потока
        sleep(0.001)


# Настройка потока приема сообщений
threading_in = threading.Thread(target=work_in)
threading_in.daemon = True
threading_in.start()


# Главная функция запускаемая в цикле
def main():
    global list_in
    global busy_in

    if len(list_in) > 0:
        while busy_in:
            sleep(0.001)
        busy_in = 1
        str_in = list_in.pop()
        busy_in = 0

        # print(str_in, 'from main func')
        display_msg(str_in)

    root.after(20, main)  # after(время_задержки_между_повторениями_в_мс, имя_функции)


def registr():
    # global server_address

    def registr_f():
        serv_addr = reg_name_entry.get()

    registr_window = Toplevel(root)
    registr_window.title('Окно регистрации')
    reg_server_address_entry = Entry(registr_window)
    reg_server_address_entry.pack()
    reg_name_entry = Entry(registr_window)
    reg_name_entry.pack()
    reg_btn = Button(text='Зарегистрироваться!', command=registr_f)
    reg_btn.pack()


# Создание интерфейса
class messenger:
    def __init__(self, root):
        self.root = root
        self.root.geometry('275x400+200+200')
        self.root.title('messenger v0.6')

        self.menu = Menu(self.root)
        self.root.config(menu=self.menu)

        self.fileMenu = Menu(self.menu, tearoff=0)
        self.menu.add_cascade(label='File', menu=self.fileMenu)

        self.fileMenu.add_command(label='Очистка истории сообщений', command=self.clear_history)
        self.fileMenu.add_command(label='Выход', command=exit)
        self.menu.add_command(label='Tools', command=self.tools_window_f)
        self.menu.add_command(label='Options', command=self.options_window_f)

        self.frame_text = Frame(root,)
        self.frame_text.place(x=5, y=5, relwidth=1, relheight=1, width=-5, height=-30)
        self.text = Text(self.frame_text, state=DISABLED)
        self.text.place(x=0, y=0, relwidth=1, relheight=1, width=-60, height=0)

        self.btn1 = Button(self.frame_text, text=1, width=7, height=3)
        self.btn1.pack(anchor=NE, pady=5)
        self.btn2 = Button(self.frame_text, text=2, width=7, height=3)
        self.btn2.pack(anchor=NE, pady=5)
        self.btn3 = Button(self.frame_text, text=3, width=7, height=3)
        self.btn3.pack(anchor=NE, pady=5)

        self.frame_entry = Frame(root,)
        self.frame_entry.pack(side=BOTTOM, fill=X)

        self.entry = Entry(self.frame_entry)
        self.entry.pack(side=LEFT, fill=X, expand=1)

        self.send_btn = Button(self.frame_entry, text='send', command=lambda: self.send_2_server(0))
        self.send_btn.pack(side=RIGHT, padx=5)

        self.entry.bind('<Return>', self.send_2_server)
        self.btn1.config(state=DISABLED)
        self.btn2.config(state=DISABLED)
        self.btn3.config(state=DISABLED)

        self.nick_name = ''

    def nick_name_f(self):
        dialog = simpledialog.askstring("Регистрация", "Введите имя")
        if dialog:
            self.nick_name = dialog

    def options_window_f(self):
        self.options_window = Toplevel()
        self.options_window.title('options')

    def tools_window_f(self):
        self.tools_window = Toplevel()
        self.tools_window.title('tools')

    def clear_history(self):
        self.text.config(state=NORMAL)
        self.text.delete(0.0, END)
        self.text.config(state=DISABLED)

    def send_2_server(self, event):
        if not self.nick_name:
            self.nick_name_f()
            if not self.nick_name:
                return 0
        msg = self.entry.get()
        self.entry.delete(0, END)
        # bin_msg = msg.encode('utf-8')             !!!!!!!!!!!!!!!
        if msg:
            bin_msg = bytes(msg, 'UTF-8')

            server_msg = ''

            client = socket(AF_INET, SOCK_STREAM)
            try:
                client.connect(server_address)
                client.sendall(bin_msg)
                data = client.recv(1024)
                server_msg = data.decode('utf-8')
                print(server_msg)
            except:
                server_msg = 'not connected'
                print('not connected')
            finally:
                client.close()

            self.text.config(state=NORMAL)
            self.text.insert(END, server_msg + '\n\n')
            self.text.config(state=DISABLED)

    def display_msg(self, msg):
        self.text.config(state=NORMAL)
        self.text.insert(END, msg + '\n')
        self.text.yview(END)
        self.text.config(state=DISABLED)

    def send_msg(self, event):
        self.text.config(state=NORMAL)
        self.msg = name + ':\n     ' + self.entry.get() + '\n\n'
        # text.highlight_pattern(name, "red")
        self.text.insert(END, self.msg)
        self.text.config(state=DISABLED)
        self.entry.delete(0, END)


# root = Tk()
# root.geometry('275x400+200+200')
# root.title('messenger v0.1')
#
# menu = Menu(root)
# root.config(menu=menu)
#
# fileMenu = Menu(menu, tearoff=0)
# menu.add_cascade(label='File', menu=fileMenu)
#
# fileMenu.add_command(label='Очистка истории сообщений', command=clear_history)
# fileMenu.add_command(label='Выход', command=exit)
# menu.add_command(label='Tools', command=tools_window_f)
# menu.add_command(label='Options', command=options_window_f)
#
# frame_text = Frame(root, )
# frame_text.place(x=5, y=5, relwidth=1, relheight=1, width=-5, height=-30)
# text = Text(frame_text, state=DISABLED)
# text.place(x=0, y=0, relwidth=1, relheight=1, width=-60, height=0)
#
# btn1 = Button(frame_text, text=1, width=7, height=3)
# btn1.pack(anchor=NE, pady=5)
# btn2 = Button(frame_text, text=2, width=7, height=3)
# btn2.pack(anchor=NE, pady=5)
# btn3 = Button(frame_text, text=3, width=7, height=3)
# btn3.pack(anchor=NE, pady=5)
#
# frame_entry = Frame(root, )
# frame_entry.pack(side=BOTTOM, fill=X)
#
# entry = Entry(frame_entry)
# entry.pack(side=LEFT, fill=X, expand=1)
#
# btn = Button(frame_entry, text='send', command=lambda: send_2_server(0))
# btn.pack(side=RIGHT, padx=5)
#
# entry.bind('<Return>', send_2_server)
#
# btn1.config(state=DISABLED)
# btn2.config(state=DISABLED)
# btn3.config(state=DISABLED)

root = Tk()
window = messenger(root)

main()

window.nick_name_f()

window.root.mainloop()



chat_UDPver_0.1.py
# чат - мессенджер
# структура сообщения: msg_nick, dest_nick, msg_data
# структура принятого сообщения: ip, msg_nick, dest_nick, msg_data, mst_time

# interface libs
import tkinter as tk
from tkinter import font
from tkinter import ttk
from tkinter import simpledialog

import socket as sock
import threading as thread
from time import *

import minimumTFTP
from tkinter import filedialog
from tkinter import messagebox
import os

# interface colors
root_color = 'cyan'
sel_color = 'red'

# sizes
btn_width = 14
text_width = 400

list_partners = [('all', '255.255.255.255', -1), ('self', '127.0.0.1', -1), ('hypercube', '192.168.1.150', -1)]
partner_timeout = 3000

# UDP socket for receive messages
HOST_IN = ''
PORT_IN = 4000
BUF_SIZE = 1024
SOCK_ADDR_IN = (HOST_IN, PORT_IN)
uServSock = sock.socket(sock.AF_INET, sock.SOCK_DGRAM)
uServSock.bind(SOCK_ADDR_IN)

# очередь для приема сообщений
list_in = []
busy_in = 0
main_tau = 30  # период цикла главной функции`

# UDP socket for transmit messages
HOST_OUT = '127.0.0.1'
PORT_OUT = 4000
# BUF_SIZE = 1024
SOCK_ADDR_OUT = (HOST_OUT, PORT_OUT)
uClientSock = sock.socket(sock.AF_INET, sock.SOCK_DGRAM, sock.IPPROTO_UDP)
uClientSock.setsockopt(sock.SOL_SOCKET, sock.SO_BROADCAST, 1)


# functions

def work_in():
    global list_in
    global busy_in
    while True:
        data, address = uServSock.recvfrom(BUF_SIZE)
        local_data = data.decode('cp1251')
        local_data = local_data.strip()

        msg_time = str(get_timems())
        str_data = address[0] + '|' + local_data + '|' + msg_time  # IP + сообщение

        while busy_in:
            sleep(0.001)
        busy_in = 2
        list_in.append(str_data)
        busy_in = 0

        sleep(0.001)
    uServSock.close()


def send_msg(msg, ip):
    print(ip)
    st = msg.encode('cp1251')
    sock_addr = (ip, PORT_OUT)
    uClientSock.sendto(st, sock_addr)


def get_timems():
    return int(time() * 1000)


def work_in_tftp():
    cud_dir = os.getcwd()
    tftpServer = minimumTFTP.Server(cud_dir)
    tftpServer.run()
    while True:
        sleep(0.001)


# create interFace
root = tk.Tk()
dFont = font.Font(family='helvetica', size=12)
style = ttk.Style()

root.configure(background=root_color)
style.configure('.', font=dFont, background=root_color, foreground='black')

ttk.Label(root, text='Simple UDP chat (ver0.1)').grid(row=0, column=0, pady=5)

# left panel. For text messages
panel_left = tk.Frame(root)
panel_left.grid(row=2, column=0, columnspan=6, rowspan=10, pady=10, padx=10)

# left panel. For show messages (main)
panel_msg = tk.Frame(panel_left, width=700)
panel_msg.grid(row=2, column=0)

# text space. For show msgs
text_bx_msg = tk.Text(panel_msg, height=30, wrap=tk.WORD, font=dFont)
text_bx_msg.pack(side='left', fill='both', expand=1)

scroll_bar_msg = tk.Scrollbar(panel_msg)
scroll_bar_msg['command'] = text_bx_msg.yview
text_bx_msg['yscrollcommand'] = scroll_bar_msg.set
scroll_bar_msg.pack(side='right', fill='y')

panel_send = tk.Frame(panel_left)
panel_send.grid(row=3, column=0, pady=5)

tbx_send = tk.Text(panel_send, height=3, wrap=tk.WORD, font=dFont)
tbx_send.pack(side='left', fill='both', expand=1)

scroll_bar_send = tk.Scrollbar(panel_send)
scroll_bar_send['command'] = tbx_send.yview
tbx_send['yscrollcommand'] = scroll_bar_send.set
scroll_bar_send.pack(side='right', fill='y')


# bind on <Enter> pressed
def function_tbx_send(event):
    my_nick = var_nick.get()
    if not my_nick:
        my_nick = simpledialog.askstring('Регистрация', 'Введите ваш ник')
        var_nick.set(my_nick)

    partner_nick = var_partner.get()
    text_send = tbx_send.get(1.0, tk.END).strip()

    if len(text_send) > 0:
        send_msg(my_nick + '|' + partner_nick + '|' + text_send, HOST_OUT)
        text_out = '\n( >>> ' + partner_nick + '  ):    ' + text_send
        text_bx_msg.insert(tk.END, text_out)
        tbx_send.delete(1.0, tk.END)


tbx_send.bind("<Return>", function_tbx_send)

# right panel
panel_right = tk.Frame(root, background=root_color)
panel_right.grid(row=0, column=8, rowspan=20)

# nick space
ttk.Label(panel_right, text='Ваше имя:').grid(row=0, column=0, sticky=tk.W, padx=5)
var_nick = tk.StringVar()
var_nick.set('')
edit_nick = ttk.Entry(panel_right, width=btn_width, textvariable=var_nick, font=dFont)
edit_nick.grid(row=1, column=0, padx=5, pady=5, sticky=tk.W)

# partner nick space
ttk.Label(panel_right, text='Партнер:').grid(row=2, column=0, sticky=tk.W, pady=5)
var_partner = tk.StringVar()
var_partner.set('')
edit_partner = ttk.Entry(panel_right, width=btn_width, textvariable=var_partner, font=dFont)
edit_partner.grid(row=3, column=0, padx=5, pady=5, sticky=tk.W)

# partners list panel
ttk.Label(panel_right, text='партнеры').grid(row=5, column=0, padx=5, pady=5)
panel_partners = tk.Frame(panel_right)
panel_partners.grid(row=6, column=0, padx=5, pady=5, sticky=tk.N)

listbox_partners = tk.Listbox(panel_partners, width=btn_width, height=16, font=dFont)
listbox_partners.pack(side='left', fill='y')
scrollbar_partners = tk.Scrollbar(panel_partners, orient='vertical')
scrollbar_partners.pack(side='right', fill='y')

scrollbar_partners.config(command=listbox_partners.yview)
listbox_partners.config(yscrollcommand=scrollbar_partners.set)


def func_set_partner(event):
    '''
    binding the double clicking at partner in listbox of partners

    :param event: None
    :return: None
    '''

    global HOST_OUT
    index = listbox_partners.curselection()
    partner = listbox_partners.get(index)
    var_partner.set(partner)
    # print(index, list_partners)
    HOST_OUT = list_partners[index[0]][1]
    # print(HOST_OUT)


listbox_partners.bind("<Double-1>", func_set_partner)


# button for sending a files
def send_file():
    if HOST_OUT == '255.255.255.255':
        msgBox = messagebox.showwarning("Error!!!!!", "U can send file only to a specific partner")
        return

    # select file in dialog window
    file_name = filedialog.askopenfilename(initialdir='.', title='выбыры какой фаел отправить',
                                           filetypes=(('all files', '*.*'), ('txt files', '*.txt')))
    if len(file_name) < 1: return
    delim_pos = file_name.rfind('/')    # получаем путь к файлу и имя файла
    path = file_name[:delim_pos]
    file_name = file_name[delim_pos+1:]
    print(path, file_name)

    tftpClient = minimumTFTP.Client(HOST_OUT, path, file_name)
    tftpClient.put()


ttk.Button(panel_right, text='send file', width=btn_width,
           command=send_file).grid(row=7, column=0, sticky=tk.W, padx=5, pady=5)

# when starting soft:
cur_time = get_timems()
list_partners[1] = ("piter", "192.168.1.200", cur_time)
# load partners list
for partner in list_partners:
    listbox_partners.insert(tk.END, partner[0])

# select first partner
listbox_partners.selection_set(0)
partner = listbox_partners.get(0)
var_partner.set(partner)
HOST_OUT = list_partners[0][1]

# create and start threading
tr_in = thread.Thread(target=work_in)
tr_in.daemon = True
tr_in.start()

# get start time
time_stamp = get_timems()

# create and start geting files thread
tr_in_tftp = thread.Thread(target=work_in_tftp)
tr_in_tftp.daemon = True
tr_in_tftp.start()


def main():
    global list_in
    global busy_in
    global time_stamp
    global HOST_OUT

    flag_partnerchange = 0  # сбросить признак изменения партнера
    my_nick = var_nick.get()

    if len(list_in) > 0:
        # carefully get msg from thread
        while busy_in:
            sleep(0.001)
        busy_in = 2
        msg_in = list_in.pop(0)
        busy_in = 0
        # put it in text space
        list_msg = msg_in.split('|')
        msg_ip, msg_nick, dest_nick, msg_data, msg_time = list_msg
        print(msg_ip, msg_nick, dest_nick, msg_data, msg_time)
        msg_time = int(msg_time)

        if msg_nick != my_nick:

            if msg_data != "*":
                text_in = '\n(  ' + msg_nick + '  >>>>   ):    ' + msg_data
                text_bx_msg.insert(tk.END, text_in)

            # find partner in list, using IP
            select_number = -1
            for num, partn in enumerate(list_partners):
                if partn[1] == msg_in and partn[0] == msg_nick:
                    list_partners[num] = (partn[0], partn[1], msg_time)
                    select_number = num
                    break

            # add partner if not find
            if select_number < 0:
                list_partners.append((msg_nick, msg_ip, msg_time))
                flag_partnerchange = 1

    # check time, and delete "old" accounts
    cur_time = get_timems()
    for num in range(len(list_partners) -1, -1, -1):
        if list_partners[num][2] == -1:  # do not delete "all"
            continue
        if cur_time > list_partners[num][2] + 3 * partner_timeout:
            del list_partners[num]
            flag_partnerchange = 1

    # show changes in partner list
    if flag_partnerchange == 1:
        listbox_partners.delete(0, tk.END)
        for partner in list_partners:
            listbox_partners.insert(tk.END, partner[0])

        listbox_partners.selection_set(0)
        partner = listbox_partners.get(0)
        var_partner.set(partner)
        HOST_OUT = '255.255.255.255'

    # send to everyone empty msg
    if cur_time - time_stamp >= partner_timeout:
        time_stamp = cur_time

        if not my_nick:
            my_nick = simpledialog.askstring('Регистрация', 'Введите ваш ник')
            var_nick.set(my_nick)

        msg = my_nick + '|all|%*%'
        send_msg(msg, '255.255.255.255')
        sleep(0.005)

    # text_in = '\n(  ' + msg_nick + '  >>>>  ):    ' + msg_data
    # tbx_msg.insert(tk.END, text_in)

    root.after(main_tau, main)


main()

root.mainloop()

uClientSock.close()
